File: namzy-cpp/src/mangle.cpp
// this_file: namzy-cpp/src/mangle.cpp
#include "mangle.h"
#include "wordlist.h"
#include <QVector>

namespace {
constexpr int kMaxLen = 12;
constexpr int kMaxTries = 8;

bool hasTripleLetter(const QString& s) {
    for (int i = 2; i < s.size(); ++i) {
        if (s.at(i) == s.at(i - 1) && s.at(i) == s.at(i - 2)) return true;
    }
    return false;
}

bool junctionUgly(const QString& compound, int junction) {
    const int start = std::max(0, junction - 2);
    const int end = std::min(compound.size(), junction + 2);
    const QString win = compound.mid(start, end - start);
    for (const QString& seam : namzy::badSeams()) {
        if (win.contains(seam)) return true;
    }
    return false;
}

QString capitalize(const QString& s) {
    if (s.isEmpty()) return s;
    return QString(s.at(0).toUpper()) + s.mid(1);
}
} // namespace

QString applyRotation(const QString& s, QRandomGenerator& rng) {
    struct Match { int i; QString src; QString dst; };
    QVector<Match> matches;
    const auto& rules = namzy::rotations();
    for (int i = 0; i < s.size(); ++i) {
        for (const auto& r : rules) {
            const QString& src = r.first;
            if (i + src.size() <= s.size() && s.mid(i, src.size()) == src) {
                matches.append(Match{ i, src, r.second });
            }
        }
    }
    if (matches.isEmpty()) return s;
    const Match& m = matches.at(static_cast<int>(rng.bounded(static_cast<quint32>(matches.size()))));
    return s.left(m.i) + m.dst + s.mid(m.i + m.src.size());
}

QString buildName(QRandomGenerator& rng) {
    const QStringList& stems = namzy::stems();
    const int n = stems.size();
    QString best;
    for (int t = 0; t < kMaxTries; ++t) {
        const QString a = stems.at(static_cast<int>(rng.bounded(static_cast<quint32>(n))));
        const QString b = stems.at(static_cast<int>(rng.bounded(static_cast<quint32>(n))));
        const QString compound = a + b;
        if (compound.size() > kMaxLen || hasTripleLetter(compound) || junctionUgly(compound, a.size())) {
            if (best.isEmpty()) best = compound;
            continue;
        }
        return capitalize(applyRotation(compound, rng));
    }
    return capitalize(applyRotation(best, rng));
}

File: namzy-rs/src/mangle.rs
//! this_file: src/mangle.rs

use crate::wordlist::{BAD_SEAMS, ROTATIONS, STEMS};

const MAX_LEN: usize = 12;
const MAX_TRIES: usize = 8;

pub struct Mulberry32 {
    state: u32,
}

impl Mulberry32 {
    pub fn new(seed: u64) -> Self {
        let state = ((seed ^ (seed >> 32)) as u32).wrapping_add(1);
        Self { state }
    }

    pub fn next_u32(&mut self) -> u32 {
        self.state = self.state.wrapping_add(0x6D2B79F5);
        let mut z = self.state;
        z = (z ^ (z >> 15)).wrapping_mul(z | 1);
        z ^= z.wrapping_add((z ^ (z >> 7)).wrapping_mul(z | 61));
        z ^ (z >> 14)
    }

    pub fn range(&mut self, n: usize) -> usize {
        if n == 0 {
            return 0;
        }
        (self.next_u32() as usize) % n
    }
}

fn has_triple_letter(s: &str) -> bool {
    let bytes = s.as_bytes();
    for i in 2..bytes.len() {
        if bytes[i] == bytes[i - 1] && bytes[i] == bytes[i - 2] {
            return true;
        }
    }
    false
}

fn junction_ugly(compound: &str, junction: usize) -> bool {
    let start = junction.saturating_sub(2);
    let end = (junction + 2).min(compound.len());
    let win = &compound[start..end];
    BAD_SEAMS.iter().any(|seam| win.contains(seam))
}

pub fn apply_rotation(s: &str, rng: &mut Mulberry32) -> String {
    let mut matches: Vec<(usize, &'static str, &'static str)> = Vec::new();
    let bytes = s.as_bytes();
    for i in 0..bytes.len() {
        for (src, dst) in ROTATIONS {
            let src_bytes = src.as_bytes();
            if i + src_bytes.len() <= bytes.len() && &bytes[i..i + src_bytes.len()] == src_bytes {
                matches.push((i, *src, *dst));
            }
        }
    }
    if matches.is_empty() {
        return s.to_string();
    }
    let (i, src, dst) = matches[rng.range(matches.len())];
    let mut out = String::with_capacity(s.len() + dst.len());
    out.push_str(&s[..i]);
    out.push_str(dst);
    out.push_str(&s[i + src.len()..]);
    out
}

fn capitalize(s: &str) -> String {
    let mut c = s.chars();
    match c.next() {
        None => String::new(),
        Some(f) => f.to_uppercase().collect::<String>() + c.as_str(),
    }
}

pub fn build_name(rng: &mut Mulberry32) -> String {
    let mut best = String::new();
    for _ in 0..MAX_TRIES {
        let a = STEMS[rng.range(STEMS.len())];
        let b = STEMS[rng.range(STEMS.len())];
        let compound = format!("{}{}", a, b);
        if compound.len() > MAX_LEN
            || has_triple_letter(&compound)
            || junction_ugly(&compound, a.len())
        {
            if best.is_empty() {
                best = compound;
            }
            continue;
        }
        return capitalize(&apply_rotation(&compound, rng));
    }
    capitalize(&apply_rotation(&best, rng))
}
